### ISSUE TITLE
RenderOpacity sometimes omits its layer during animations which breaks caching performance

### ISSUE THREAD

**Issue by flar** [flutter/flutter#58425]
The RenderOpacity object will not emit a layer when the opacity is 0 or 255. If that is the static opacity then it might make sense, but when it is the value during an animation, this causes the layer tree to drastically change configuration on each frame.

While this could potentially result in a hit if the tree needs to be constantly reconstructed, an even bigger hit happens because the change of identity of its own layer as it moves from a partial value to either 0 or 255 and back again means that the cached information at the engine level needs to be discarded and then re-rendered.

The layer should absolutely be left intact even for opacity values of 0 or 255 during an animation.

**Comment by goderbauer** [comment]
At the `RenderOpacity` level we do not know that the opacity change is part of an animation. We could change it to always push the layer no matter what the opacity value is. I am wondering whether having the layer there all the time impacts performance in the static case where opacity is fixed at 0 or 255.

Is there a benchmark that shows that adding/removing the layer in an animation is affecting performance badly?

**Comment by flar** [comment]
I can write a benchmark that shows the difference, but it is subtle and the difference will often slip through the numbers we record for benchmarks. The effect will be a slower frame every Nth frame as the animation goes through its range which is more of a jank problem than a benchmark statistic problem.

But, the performance turns out to not be the most noticeable issue for developers. Instead we have situations like https://github.com/flutter/flutter/issues/60966 (A BackdropFilter inside an Opacity) where the issue is that the semantics change when the Opacity widget asserts itself with a layer. A BackdropFilter cannot blur a background when it has a (SaveLayer-type) layer (such as Opacity layers) between it and the background. Since the Opacity does not insert a layer when its value is 1.0 the developer sets up this situation and it looks fine. Then when they trigger the fade animation suddenly the rendering changes.

When the opacity layer is 1.0 I don't know how much is saved by omitting the layer, but in terms of whether or not to insert the SaveLayer that changes semantics, that optimization could also be made at the engine layer so the rendering benefit doesn't have to be made at the widget/render object level.

When the opacity layer is 0.0, though, we could potentially save processing the entire child tree at the framework level which will have impact that depends on the complexity of the child, so it is harder to assess the performance impact of severing the child in further processing. Also, when the value is 0.0 then nothing of the child should show which means that the change in behavior is probably invisible anyway (I'd have to dig further to see if I could construct an example with non-default blending modes that might show the difference, but I don't think it is possible as I don't think we use them in a way that could affect how an Opacity layer blends into the scene behind it).

**Comment by goderbauer** [comment]
We currently do stop processing the child tree at the framework level when opacity is 0.0. Based on your previous post, can we continue to do that (and not push a layer) for the 0.0 case? And maybe just always push a layer for the 255 case?

**Comment by flar** [comment]
I'm still looking into it, unfortunately I have a few other things on my plate right now. The outstanding questions are:

- Can an app tell the difference between a missing tree and one rendered via a SaveLayer with an opacity of 0.0? (I think yes, but I need to investigate blending modes)

- Can we modify the BackdropFilter implementation to erase the behavior difference when the SaveLayer goes away? (For example: https://github.com/flutter/engine/pull/19631 which hasn't been fully tested or proven to work for all cases)

- Can we modify the BackdropFilter implementation to alleviate the other behavior change that happens when a SaveLayer is introduced? The change above affects whether the blur effect in the BackdropFilter obscures the original text, but another behavior change not helped by that PR is that the BackdropFilter no longer has access to anything behind the OpacityLayer when it tries to blur things. That second behavior change would take a lot to fix.

- How much do we care about this particular aspect of fading_child_animation_perf__timeline_summary? The test animates an opacity from 0.0 to 1.0 back and forth repeatedly. Is that a realistic case for apps in the wild? Does the benchmark reliably hit 0.0 and 1.0 exactly anyway or is the chance of hitting either of those cases random and non-reproducible? Do we want to include the case of transitioning to/from 0.0 and 1.0 in the benchmark?

If we want to include the layer when the opacity is 1.0, then we can do the same check in the engine layer. We can also check in the engine layer if any lower layer behaves differently when it is in the presence of a SaveLayer and be smart about when we omit the extra layer. We'd want to coordinate those 2 implementation changes.

With respect to not omitting the child tree in the case of 0.0, that would require more testing and analysis and I don't yet have a complete picture to answer that part of the question.

**Comment by flar** [comment]
We could have a separate discussion on whether or not we care to benchmark the case of animating opacity from 0 to 1 back and forth repeatedly. That benchmark, which is primarily about measuring the improvement of the partial opacity cases, could be modified to bounce between .1 and .9 to avoid the end conditions, or it could slowly animate from 0 to 1 over the entire run of the benchmark instead of ping-ponging back and forth.

**Comment by flar** [comment]
I've looked through the SaveLayer code as used by the engine OpacityLayer and it looks like the only way the save layer gets drawn back to the original surface is through the default SrcOver blending mode so 1.0 and 0.0 do behave as we expect.

**Comment by flar** [comment]
Another behavior change is that RenderObjects that don't use layers but just paint will share a Picture object between them. This sharing is disrupted when the tree traversal encounters a RenderObject that inserts a layer. So, if the Opacity(1.0) is just passing through to its children then it will allow Picture sharing from its ancestors to its children and then back again to its subsequent siblings and ancestors. If the opacity value changes and it starts to insert its layer then the sharing will stop.

There should be no visible difference when that happens (other than the opacity introduced on some of the items), but any caching of those Picture objects will also be disrupted. As such, when an Opacity widget with a value of 1.0 is in the middle of a large, mostly graphical frame, then it will share one Picture across the entire frame and that picture will be cached. If the value changes then we will have to switch to 3 different Picture objects (one for everything before the Opacity, one for the children of the Opacity, and one for everything after the Opacity). This will require recaching all of the rendering into 3 separate picture cache objects and it will only happen after rendering the first 3 frames. This would add more jank to the start of any fade animation. If the layer is always inserted then the cache of all 3 parts would be preserved and only the presence of a new SaveLayer at the rendering stage would differentiate their performance.

**Comment by flar** [comment]
Tagging @liyuqian

**Comment by liyuqian** [comment]
I agree with https://github.com/flutter/flutter/issues/58425#issuecomment-661267655 that the correctness issues are more important than the performance issues here because no users have complained about the performance difference between having or not having an additional layer, yet we've already had user complaints about the correctness of the rendering.

For correctness, my intuition is:
- It's correct to drop the whole subtree if the opacity is 0.
- It's incorrect to remove `saveLayer` if the opacity is 1. (By removing it, the 0.99-opacity rendering could be very different from 1-opacity  rendering and cause the animation to look very bad.)

**Comment by flar** [comment]
Could the translation shift slightly when using a SaveLayer? If so, then that is another discontinuity between 0.99 and 1.0.

**Comment by goderbauer** [comment]
I am no longer convinced that always pushing a layer even when opacity is 1.0 is the right thing for RenderOpacity. I am afraid that this will unnecessary slow things down for common `Opacity` use cases. The `Visibility` widget for example toggles the Opacity widget behind the scene between just 0.0 and 1.0. In that case it seems totally unnecessary (and somewhat surprising) to push a layer.

I am not sure what the best way forward is. Maybe we need to add a flag to RenderOpacity that indicates whether you want to push the layer at 1.0 or not? For backwards compatibility, that flag should be false. The visibility widget would leave it as false, but if you want to animate the opacity (or switch it between 1.0 and non-1.0 values) you'd have to set it to true?

Having said that, for the RenderAnimatedOpacityMixin we should always push a layer even when opacity is 1.0 since the whole point of that widget is animation.

What do people think?

**Comment by flar** [comment]
Perhaps we need a Toggle widget to replace what Visibility is doing?

Also, your concern is that always pushing a layer "might" affect performance. We should probably do some benchmarks to check, but keep in mind that the engine can also control the performance impact by how the engine OpacityLayer is implemented. Right now, it has no ability to do that because it never sees the layer in the first place when it gets omitted during the build process.

**Comment by flar** [comment]
Visibility wrapping the child with an Opacity widget when visible is true is already its own performance issue.

**Comment by flar** [comment]
> I am no longer convinced that always pushing a layer even when opacity is 1.0 is the right thing for RenderOpacity. I am afraid that this will unnecessary slow things down for common `Opacity` use cases. The `Visibility` widget for example toggles the Opacity widget behind the scene between just 0.0 and 1.0. In that case it seems totally unnecessary (and somewhat surprising) to push a layer.
> 
> I am not sure what the best way forward is. Maybe we need to add a flag to RenderOpacity that indicates whether you want to push the layer at 1.0 or not? For backwards compatibility, that flag should be false. The visibility widget would leave it as false, but if you want to animate the opacity (or switch it between 1.0 and non-1.0 values) you'd have to set it to true?
> 
> Having said that, for the RenderAnimatedOpacityMixin we should always push a layer even when opacity is 1.0 since the whole point of that widget is animation.
> 
> What do people think?

It looks like you pulled your PR. Any thoughts on the Visibility widget so that Opacity is only used for actual translucency cases? You mentioned that the Animated mixin should probably always insert the layer, is there a new PR for that?

**Comment by flar** [comment]
There may be a simpler solution than keeping the widget when it isn't needed. Basically just ensure that the Picture objects are not glued together with or across this widget by careful manipulation of the current recording - with the potential addition of adding a RepaintBoundary mechanism to prevent opacity changes from forcing repaints of the surrounding content.

See https://github.com/flutter/flutter/pull/83145#issuecomment-858182715

**Comment by flar** [comment]
Note that a customer noticed a difference in rendering when alpha is 255 due to the missing layer as shown in https://github.com/flutter/flutter/issues/93378

If the child is using nonstandard blend modes, then the presence of a saveLayer directly impacts what it looks like.
